### PR TITLE
Support function and operation types without parentheses

### LINKED ISSUE #788: Support function and operation types without parentheses
**Is your feature request related to a problem? Please describe.**
The parentheses required around function and operation types is excessive. For example:

```qsharp
function Apply1(f : (Int -> (Int -> Int))) : (Int -> Int) {
    return f(1);
}

function ApplyAdjoint(op : (Qubit => Unit is Adj)) : (Qubit => Unit is Adj) {
    return Adjoint op;
}
```

**Describe the solution you'd like**
I would prefer to write:

```qsharp
function Apply1(f : Int -> Int -> Int) : Int -> Int {
    return f(1);
}

function ApplyAdjoint(op : Qubit => Unit is Adj) : Qubit => Unit is Adj {
    return Adjoint op;
}
```

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by samarsha**
This PR changes the syntax for function and operation types. Parentheses are optional in most cases. You can write:

```qsharp
function Foo(f : Int -> Int -> Int) : Int -> Int {
    return f(1);
}
```

instead of

```qsharp
function Foo(f : (Int -> (Int -> Int))) : (Int -> Int) {
    return f(1);
}
```

Parentheses are required when a callable type is used as the item type of an array. For example, `(Int -> Int)[]`, `(Qubit => Unit is Adj)[]`, or `new (Qubit => Unit is Adj)[3]`.

Two breaking changes were made, which we expect to have minimal impact on existing code:

1. Remove support for using a colon to start operation characteristics, e.g. `Qubit => Unit : Adjoint`. This syntax has been deprecated for a while.
2. Remove support for parentheses around the arrow and before the characteristics in operation types, e.g. `((Qubit => Unit) is Adj)`. This must be written as `(Qubit => Unit is Adj)` or `Qubit => Unit is Adj`.

The reason for the second change is to distinguish between these two cases:

```qsharp
operation Foo() : Qubit => Unit is Adj {
    // ...
}

operation Bar() : (Qubit => Unit) is Adj {
    // ...
}
```

This PR defines the meaning of this example to be:

1. `Foo` returns a value of type `Qubit => Unit is Adj`. `Foo` is not adjointable.
2. `Bar` returns a value of type `Qubit => Unit`. `Bar` is adjointable. This is a type error, because adjointable or controllable operations must return `Unit`, but allowing this to be parsed keeps the possibility open for future language development related to this.

Closes #788.

### CODE REVIEW

**Comment by ScottCarda-MS** [comment]
I was wondering if we could have a more formal design review for this feature (#788). (Or maybe we already did and I missed it?) There are some questions I have going into this that I would want to talk about over a meeting and that would give people the opportunity to chime in with their questions and comments.

**Comment by samarsha** [comment]
@ScottCarda-MS Could you post your comments/questions here or in issue #788?

**Comment by ScottCarda-MS** [comment]
> @ScottCarda-MS Could you post your comments/questions here or in issue #788?

Is that to say that you will not be doing a formal design review for this?

**Comment by samarsha** [comment]
@ScottCarda-MS It would be helpful to understand specifically what your concerns are with this change. While it is technically a change to the language that could go through the Q# proposal process, @bettinaheim and I feel that since it's a relatively minor change to parsing, it could go through normal PR review for the compiler instead.

If you post your concerns, we can see if we can address them here or if we need more discussion in a language proposal.

**Comment by bettinaheim** [comment]
> @ScottCarda-MS It would be helpful to understand specifically what your concerns are with this change. While it is technically a change to the language that could go through the Q# proposal process, @bettinaheim and I feel that since it's a relatively minor change to parsing, it could go through normal PR review for the compiler instead.
> 
> If you post your concerns, we can see if we can address them here or if we need more discussion in a language proposal.

I agree. Since this change merely allows to omit some parenthesis that were previously required, I don't think that warrants a full proposal. The usage of e.g. `(Int -> Int)` will continue to work. Additionally, `Int -> Int` is now supported with this PR as well.

File: src/QsCompiler/TextProcessor/QsTypeParsing.fs [hunk 1]
```diff
@@ -171,73 +172,89 @@ let private operationType =
...
 /// The tuple must consist of at least one tuple item.
 let internal tupleType =
     let buildTupleType (items, range: Range) = (TupleType items, range) |> QsType.New
     buildTuple qsType buildTupleType ErrorCode.InvalidType ErrorCode.MissingType invalidType
 
+/// <summary>
+/// Validates that a parsed Q# type follows additional rules that prevent confusing syntax.
+/// </summary>
+/// <param name="isArrayItem">Whether the given type is part of an array postfix bracket type, "T[]" or "T[n]".</param>
+let internal validateTypeSyntax isArrayItem { Type = kind; Range = range } =
```
Comment (bettinaheim) [inline src/QsCompiler/TextProcessor/QsTypeParsing.fs:215]: I see you went with disallowing this after all. :)

**Review by bettinaheim — APPROVED** [review]

File: src/QsCompiler/TextProcessor/QsTypeParsing.fs (hunk 1 above)
Comment (samarsha) [reply src/QsCompiler/TextProcessor/QsTypeParsing.fs:215]: Originally, I just disallowed `Qubit => Unit is Adj[]` as we discussed, but then I realized the same function could be used to check `new Int -> Int[3]` as well, so it seemed like a good idea.

File: src/QsCompiler/TextProcessor/QsTypeParsing.fs [hunk 2]
```diff
@@ -171,73 +172,89 @@ let private operationType =
...
-        let withoutInnerBrackets = optTupleBrackets (inAndOutputType .>>. characteristics)
-        withInnerBrackets <|> withoutInnerBrackets
-
-    let deprecatedCharacteristics =
-        let colonWithWarning = buildWarning (getEmptyRange .>> colon) WarningCode.DeprecatedOpCharacteristicsIntro
-
-        attempt (colonWithWarning >>. characteristics .>> notFollowedBy (comma >>. quantumFunctor))
-        <|> (qsCharacteristics.parse |>> (fun r -> r.Start) <|> (getPosition .>> colon) >>= functorSupport)
-
+    let opTypeWith characteristics = inAndOutputType .>>. characteristics
```
[the remark this answers is not in the document]
Comment (samarsha) [reply src/QsCompiler/TextProcessor/QsTypeParsing.fs]: I think to fully factor it it's more like:

```fsharp
inAndOutputType
.>>. (attempt characteristics <|> deprecatedCharacteristics <|>% Characteristics.New(EmptySet, Null))
```

right?

[the remark this answers is not in the document]
Comment (bettinaheim) [reply src/QsCompiler/TextProcessor/QsTypeParsing.fs]: I am not sure I understand? Support for the deprecated characteristics (e.g. `: Adjoint`) is removed with this PR, so I was thinking something like `opt characteristics |>> function | Some char -> char | None -> Characteristics.New(EmptySet, Null))`

[the remark this answers is not in the document]
Comment (samarsha) [reply src/QsCompiler/TextProcessor/QsTypeParsing.fs]: > Support for the deprecated characteristics (e.g. `: Adjoint`) is removed with this PR

Yes, the colon syntax is removed, but there is another kind of deprecated characteristics syntax that I left in: `is Adj, Ctl` is equivalent to `is Adj + Ctl`.

> ```fsharp
> opt characteristics |>> function | Some char -> char | None -> Characteristics.New(EmptySet, Null))
> ```

Note that `opt p` is `(p |>> Some) <|>% None`, so this is equivalent to `characteristics <|>% Characteristics.New(EmptySet, Null))`.

[the remark this answers is not in the document]
Comment (bettinaheim) [reply src/QsCompiler/TextProcessor/QsTypeParsing.fs]: All of the deprecated patterns you mention were deprecated at the same time, so I think it is fine to remove all of them.

**Pull request merged**